backend/server/src/services/tournamentService.ts

```diff
@@ -91,5 +141,169 @@ export class TournamentService {
 
     tournament.players.push(player.id);
     await tournament.save();
+
+    if (tournament.players.length > 1) {
+      const newMatchIds = await this.generateTournamentSchedule(
+        tournament,
+        player.id
+      );
+      if (newMatchIds.length !== 0) {
+        tournament.matchSchedule.push(...newMatchIds);
+        await tournament.save();
+      }
+    }
+    return await tournament.toObject();
+  }
+
+  public async addMatchToTournament(
+    tournamentId: string,
+    unsavedMatch: UnsavedMatch
+  ): Promise<Tournament> {
+    const tournament = await TournamentModel.findById(tournamentId).exec();
+    if (tournament === null || tournament === undefined) {
+      throw new NotFoundError({
+        message: "Tournament not found"
+      });
+    }
+
+    const currentDate = new Date();
+    const startDate = new Date(tournament.startDate);
+    if (currentDate > startDate) {
+      throw new BadRequestError({
+        message: `Cannot add new players as the tournament has already started on ${startDate.toDateString()}`
+      });
+    }
+
+    // validation
+    for (let i = 0; i < unsavedMatch.players.length; i += 1) {
+      if (!tournament.players.includes(unsavedMatch.players[i].id)) {
+        const user = await UserModel.findById(
+          unsavedMatch.players[i].id
+        ).exec();
+
+        if (user === null || user === undefined) {
+          throw new NotFoundError({
+            message: "Player not found!"
+          });
+        }
+        throw new BadRequestError({
+          message: `Cannot create the match: ${user.firstName} ${user.lastName} is not registered for this tournament.`
+        });
+      }
+    }
```

Comment:
The use of the for...of statement is recommended over the for statement when iterating through iterable objects as simplifies the syntax and eliminates the need for a counter variable.